parser/openfoodfacts_taxonomy_parser/parser/parser.py

```diff
@@ -0,0 +1,199 @@
+        self.parser_logger.info("Creating 'is_child_of' links")
+        for child_link in child_links:
+            child_id = child_link["id"]
+            parent = child_link["parent_id"]
+            lc, parent_id = parent.split(":")
+            query = f""" MATCH (p:{multi_label}:ENTRY) WHERE $parent_id IN p.tags_ids_""" + lc
+            query += f"""
+                MATCH (c:{multi_label}) WHERE c.id= $child_id
+                CREATE (c)-[r:is_child_of]->(p)
+                RETURN r
+            """
+            result = self.session.run(query, parent_id=parent_id, child_id=child_id)
+            if not result.value():
+                self.parser_logger.warning(
+                    f"parent not found for child {child_id} with parent {parent_id}"
+                )
+
+    def _create_fulltext_index(self, taxonomy_name: str, branch_name: str):
+        """Create indexes for search"""
+        project_name = self._get_project_name(taxonomy_name, branch_name)
+        query = (
+            f"""CREATE FULLTEXT INDEX {project_name+'_SearchIds'} IF NOT EXISTS
+            FOR (n:{project_name}) ON EACH [n.id]\n"""
+            + """
+            OPTIONS {indexConfig: {`fulltext.analyzer`: 'keyword'}}"""
+        )
+        self.session.run(query)
+
+        language_codes = [lang.alpha2 for lang in list(iso639.languages) if lang.alpha2 != ""]
+        tags_prefixed_lc = ["n.tags_" + lc for lc in language_codes]
+        tags_prefixed_lc = ", ".join(tags_prefixed_lc)
+        query = f"""CREATE FULLTEXT INDEX {project_name+'_SearchTags'} IF NOT EXISTS
+            FOR (n:{project_name}) ON EACH [{tags_prefixed_lc}]"""
+        self.session.run(query)
+
+    def _create_parsing_errors_node(self, taxonomy_name: str, branch_name: str):
+        """Create node to list parsing errors"""
+        multi_label = self._create_multi_label(taxonomy_name, branch_name)
+        query = f"""
+            CREATE (n:{multi_label}:ERRORS)
+            SET n.id = $project_name
+            SET n.branch_name = $branch_name
+            SET n.taxonomy_name = $taxonomy_name
+            SET n.created_at = datetime()
+            SET n.warnings = $warnings_list
+            SET n.errors = $errors_list
+        """
+        params = {
+            "project_name": self._get_project_name(taxonomy_name, branch_name),
+            "branch_name": branch_name,
+            "taxonomy_name": taxonomy_name,
+            "warnings_list": self.parser_logger.parsing_warnings,
+            "errors_list": self.parser_logger.parsing_errors,
+        }
+        self.session.run(query, params)
+
+    def __call__(self, filename: str, branch_name: str, taxonomy_name: str):
+        """Process the file"""
+        branch_name = normalizing(branch_name, char="_")
+        multi_label = self._create_multi_label(taxonomy_name, branch_name)
+        taxonomy_parser = TaxonomyParser()
+        taxonomy = taxonomy_parser.parse_file(filename, self.parser_logger)
```

Comment:
I would have made this a bit more composable.

For me there is the taxonomy parser that returns a Taxonomy,this is your TaxonomyParser and the taxonomy backend that takes a taxonomy and put it in Neo4J (so this class would have been Neo4jTaxonomy for example or Neo4jBackend.

Then a function (or callable class) would weaved them together. this function would be those four lines + a cll to the backend (to run the lines below). It would take the TaxonomyParser object and the Taxonomy Backend object as parameters.
It would make it clearer that you could add a PostgreSQL backend.

Reply:
It seems a bit over-engineered considering that we are currently only writing to a Neo4J database
Do you think it is essential to do this now? If we ever add support for another database, we could do the refactor at that time

Reply:
I'm ok to postpone it, though I was seeing this as a simple refactor that adds clarity.

Do as you wish :-)